cloudmarker/util.py

```diff
@@ -453,6 +455,118 @@ def pluralize(count, word, *suffixes):
+    protocols in use:
+
+    - SSL/TLS (or implicit SSL/TLS): SSL/TLS is used from the beginning
+      of the connection. This occurs typically on port 465. This is
+      enabled by default (``ssl_mode`` as ``'ssl'``).
+
+    - STARTTLS (or explicit SSL/TLS): The SMTP session begins as a
+      plaintext session. Then the client (this function in this case)
+      makes an explicit request to switch to SSL/TLS by sending the
+      ``STARTTLS`` command to the server. This occurs typically on port
+      25 or port 587. Set ``ssl_mode`` to ``'starttls'`` to enable this
+      behaviour
+
+    If ``username`` is unspecified or specified as an empty string, no
+    SMTP authentication is done. If ``username`` is specified as a
+    non-empty string, then SMTP authentication is done.
+
+    Arguments:
+        from_addr (str): Sender's email address.
+        to_addrs (list): A list of :obj:`str` objects where each
+            :obj:`str` object is a recipient's email address.
+        subject (str): Email subject.
+        content (str): Email content.
+        host (str): SMTP host.
+        port (int): SMTP port.
+        ssl_mode (str): SSL mode to use: ``'ssl'`` for SSL/TLS
+            connection (the default), ``'starttls'`` for STARTTLS, and
+            ``'disable'`` to disable SSL.
+        username (str): SMTP username.
+        password (str): SMTP password.
+        debug (int or bool): Debug level to pass to
+            :meth:`SMTP.set_debuglevel` to debug an SMTP session. Set to
+            ``0`` (the default) or ``False`` to disable debugging. Set
+            to ``1`` or ``True`` to see SMTP messages. Set to ``2`` to
+            see timestamped SMTP messages.
+    """
+    log_data = ('from_addr: {}; to_addrs: {}; subject: {}; host: {}; '
+                'port: {}; ssl_mode: {}'
+                .format(from_addr, to_addrs, subject, host, port, ssl_mode))
+    try:
+        if ssl_mode == 'ssl':
+            smtp = smtplib.SMTP_SSL(host, port)
+            smtp.set_debuglevel(debug)
+        elif ssl_mode == 'starttls':
+            smtp = smtplib.SMTP(host, port)
+            smtp.set_debuglevel(debug)
+            smtp.starttls()
+        elif ssl_mode == 'disable':
+            smtp = smtplib.SMTP(host, port)
+            smtp.set_debuglevel(debug)
+        else:
+            _log.error('Cannot send email; %s; error: %s: %s', log_data,
+                       'invalid ssl_mode', ssl_mode)
+            return
+
+        if username:
+            smtp.login(username, password)
+
+        msg = email.message.EmailMessage()
+        msg['From'] = from_addr
+        msg['To'] = ', '.join(to_addrs)
+        msg['Subject'] = subject
+        msg.set_content(content)
+
+        smtp.send_message(msg)
+        smtp.quit()
+
+        _log.info('Sent email successfully; %s', log_data)
+
+    except Exception as e:
+        _log.error('Failed to send email; %s; error: %s: %s', log_data,
+                   type(e).__name__, e)
+
+
```

Comment:
I think it would be good to keep all the exception in one `exception` package rather than adding it to the `util.py`.  The `util.py` should only be used for generic implementations which can be used all over the project by any module.

Reply:
I am confused. How is the `_log.error` statement related to keeping exceptions in an `exception` package? Can you elaborate a little more about what exactly would you like me to do?

Reply:
Please refer to line no. 570. I don't know why the line is not getting reflected in the comment.
Exceptions like `PluginError`, `PortRangeError`, `PluralizeError` should be in one `exception` package

Reply:
Could you please justify your recommendations? Why should they be in one `exception` package? What benefits does following such a practice offers?

More importantly, are those benefits so strong that they warrant breaking existing conventions? In most languages I have worked with such as C++, Java, and Python, exceptions always belong to the same namespace, package, or module, respectively. I have never seen an example where a new namespace, package, or module was created only to keep exceptions.

That makes sense too. For example, if I am parsing JSON with `json` module and I want to handle exceptions for it, I don't want to go searching the entire documentation to see what exceptions I might want to handle. I would like to type `json.` and rely on autocomplete to show me all the exceptions or just read the `json` module documentation. Similarly, we expect `smtplib` exceptions to be in the `smtplib` module. Similarly, in Java, we expect I/O exceptions to be in `java.io`, SQL exception to be in `java.sql`, and so on. In C++, for example, an I/O exception is found in `std::ios_base` namespace.